src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -38,56 +38,117 @@ size_t typelookup_service_request_nops = sizeof (typelookup_service_request_ops)
 const enum pserop typelookup_service_reply_ops[] = { XG, Xl, XQ, XK, XO, XSTOP, XSTOP };
 size_t typelookup_service_reply_nops = sizeof (typelookup_service_reply_ops) / sizeof (typelookup_service_reply_ops[0]);
 
-int ddsi_tl_meta_equal (const struct tl_meta *a, const struct tl_meta *b)
-{
-  return ddsi_typeid_equal (a->type_id, b->type_id);
-}
+typedef struct tlm_endpoints_iter {
+  const struct tl_meta *tlm;
+  uint32_t idx;
+} tlm_endpoints_iter_t;
 
-uint32_t ddsi_tl_meta_hash (const struct tl_meta *tl_meta)
+static bool tlm_endpoint_exists (struct tl_meta *tlm, const ddsi_guid_t *proxy_ep_guid)
 {
-  uint32_t h = 0;
-  h = ddsrt_mh3 (&tl_meta->type_id->hash, TYPEID_HASH_LENGTH, h);
-  return h;
+  for (uint32_t n = 0; n < tlm->proxy_endpoints.count; n++)
+  {
+    if (!memcmp (&tlm->proxy_endpoints.eps[n], proxy_ep_guid, sizeof (*proxy_ep_guid)))
+      return true;
+  }
+  return false;
 }
 
-static void ddsi_tl_meta_fini (struct tl_meta *tlm)
+static void tlm_endpoint_add (struct ddsi_domaingv *gv, struct tl_meta *tlm, const ddsi_guid_t *proxy_ep_guid)
 {
-  if (tlm->sertype != NULL)
-    ddsi_sertype_unref ((struct ddsi_sertype *) tlm->sertype);
-  ddsrt_free (tlm->type_id);
-  ddsrt_free (tlm->endpoints);
-  ddsrt_free (tlm);
+  if (proxy_ep_guid != NULL)
+  {
+    if (!tlm_endpoint_exists (tlm, proxy_ep_guid))
```

Comment:
This data structure doesn't scale. As long as systems are small, it is not an issue, and it doesn't make sense to delay merging the PR (especially considering that it defaults to being disabled) because of it, but I do think it needs to be noted.